File: 2470-removing-stars-from-a-string/removing-stars-from-a-string.java
class Solution {
    public String removeStars(String s) {
        Stack<Character> stack = new Stack();

        for(char c : s.toCharArray()){
            if(!stack.isEmpty() && c == '*'){
                stack.pop();
            } else if (c != '*') {
                stack.push(c);
            }
        }

        StringBuilder result = new StringBuilder();

        for(char c : stack){
            result.append(c);
        }

        return result.toString();
    }
}

File: 2470-removing-stars-from-a-string/removing-stars-from-a-string.scala
object Solution {
    def removeStars(s: String): String = {
        def iter(chars: List[Char], result: List[Char]): List[Char] = chars match {
            case Nil => result
            case '*' :: tail => 
                iter(tail, if result.nonEmpty then result.tail else result)
            case head :: tail =>
                iter(tail, head :: result)
        }

        iter(s.toList, Nil).reverse.mkString
    }
}
